lib/usdUfe/ufe/UsdClipboardCommands.cpp

```diff
@@ -0,0 +1,496 @@
+static_assert(!std::is_move_assignable<UsdPasteClipboardCommand>::value);
+
+UsdPasteClipboardCommand::UsdPasteClipboardCommand(
+    const Ufe::Selection&    dstParentItems,
+    const UsdClipboard::Ptr& clipboard)
+    : _clipboard(clipboard)
+{
+    for (const auto& parentItem : dstParentItems) {
+        if (const auto usdParentItem = std::dynamic_pointer_cast<UsdSceneItem>(parentItem)) {
+            _dstParentItems.push_back(usdParentItem);
+        }
+    }
+}
+
+UsdPasteClipboardCommand::UsdPasteClipboardCommand(
+    const Ufe::SceneItem::Ptr& dstParentItem,
+    const UsdClipboard::Ptr&   clipboard)
+    : _clipboard(clipboard)
+{
+    if (const auto usdItem = std::dynamic_pointer_cast<UsdSceneItem>(dstParentItem)) {
+        _dstParentItems.push_back(usdItem);
+    }
+}
+
+UsdPasteClipboardCommand::Ptr UsdPasteClipboardCommand::create(
+    const Ufe::SceneItem::Ptr& dstParentItem,
+    const UsdClipboard::Ptr&   clipboard)
+{
+    if (!dstParentItem)
+        return {};
+    return std::make_shared<UsdPasteClipboardCommand>(dstParentItem, clipboard);
+}
+
+UsdPasteClipboardCommand::Ptr UsdPasteClipboardCommand::create(
+    const Ufe::Selection&    dstParentItems,
+    const UsdClipboard::Ptr& clipboard)
+{
+    if (dstParentItems.empty())
+        return {};
+    return std::make_shared<UsdPasteClipboardCommand>(dstParentItems, clipboard);
+}
+
+void UsdPasteClipboardCommand::execute()
+{
+    UsdUndoBlock undoBlock(&_undoableItem);
+
+    // Get the Clipboard stage.
+    auto clipboardData = _clipboard->getClipboardData();
+
+    if (!clipboardData) {
+        const std::string error = "Failed to load Clipboard stage.";
+        throw std::runtime_error(error);
+    }
+
+    // Duplicate the first-level in depth items from the Clipboard stage to the destination
+    // parent item.
+    Ufe::Selection clipboardMaterials;
+    Ufe::Selection clipboardShaders;
+    Ufe::Selection clipboardPrims;
+
+    for (auto prim : clipboardData->Traverse()) {
+        // Add to the selection only the first-level in depth items.
+        if (prim.GetParent() == clipboardData->GetPseudoRoot()) {
+            Ufe::PathSegment segment(
+                "/" + prim.GetName().GetString(), UsdUfe::getUsdRunTimeId(), '/');
+
+            if (auto usdItem = UsdSceneItem::create(segment, prim)) {
+                if (PXR_NS::UsdShadeMaterial(prim)) {
+                    clipboardMaterials.append(usdItem);
+                } else if (PXR_NS::UsdShadeShader(prim) || PXR_NS::UsdShadeNodeGraph(prim)) {
+                    clipboardShaders.append(usdItem);
+                } else {
+                    clipboardPrims.append(usdItem);
+                }
+            }
+        }
+    }
+
+    if (clipboardPrims.empty() && clipboardMaterials.empty() && clipboardShaders.empty()) {
+        return; // nothing to paste
+    }
+
+    auto appendToVector = [](const auto&             itemsToAppend,
```

Comment:
Wished helper functions would be functions and not lambda in the middle of the code, but that is merely my preference.

Reply:
This was original code from LookdevX that I just brought over.

Reply:
One reason it is a lamdba is because as input it takes either a Ufe::SceneItemList or Ufe::Selection (notice the first param is auto). I could either template it or put the Ufe::SceneItemList in Ufe::Selection. @pierrebai-adsk Is that something you think should be done?